resolver-dns/src/main/java/io/netty/resolver/dns/DefaultAuthoritativeDnsServerCache.java

```diff
@@ -0,0 +1,319 @@
+
+    private final ConcurrentMap<String, Entries> resolveCache = PlatformDependent.newConcurrentHashMap();
+    private final int minTtl;
+    private final int maxTtl;
+
+    /**
+     * Create a cache that respects the TTL returned by the DNS server.
+     */
+    public DefaultAuthoritativeDnsServerCache() {
+        this(0, MAX_SUPPORTED_TTL_SECS);
+    }
+
+    /**
+     * Create a cache.
+     *
+     * @param minTtl the minimum TTL
+     * @param maxTtl the maximum TTL
+     */
+    public DefaultAuthoritativeDnsServerCache(int minTtl, int maxTtl) {
+        this.minTtl = Math.min(MAX_SUPPORTED_TTL_SECS, checkPositiveOrZero(minTtl, "minTtl"));
+        this.maxTtl = Math.min(MAX_SUPPORTED_TTL_SECS, checkPositiveOrZero(maxTtl, "maxTtl"));
+        if (minTtl > maxTtl) {
+            throw new IllegalArgumentException(
+                    "minTtl: " + minTtl + ", maxTtl: " + maxTtl + " (expected: 0 <= minTtl <= maxTtl)");
+        }
+    }
+
+    /**
+     * Returns the minimum TTL of the cached DNS resource records (in seconds).
+     *
+     * @see #maxTtl()
+     */
+    public int minTtl() {
+        return minTtl;
+    }
+
+    /**
+     * Returns the maximum TTL of the cached DNS resource records (in seconds).
+     *
+     * @see #minTtl()
+     */
+    public int maxTtl() {
+        return maxTtl;
+    }
+
+    @Override
+    public List<InetSocketAddress> get(String hostname) {
+        checkNotNull(hostname, "hostname");
+
+        Entries entries = resolveCache.get(appendDot(hostname));
+        if (entries == null) {
+            return Collections.emptyList();
+        }
+        return entries.get();
+    }
+
+    @Override
+    public void cache(String hostname, InetSocketAddress address, long originalTtl, EventLoop loop) {
+        checkNotNull(hostname, "hostname");
+        checkNotNull(address, "address");
+        checkNotNull(loop, "loop");
+        if (maxTtl == 0) {
+            return;
+        }
+        if (PlatformDependent.javaVersion() >= 7 && address.getHostString() == null) {
+            // We only cache addresses that have also a host string as we will need it later when trying to replace
+            // unresolved entries in the cache.
+            return;
+        }
+
+        final Entry e = new Entry(address);
+
+        String key = appendDot(hostname);
+        Entries entries = resolveCache.get(key);
+        if (entries == null) {
+            entries = new Entries(e);
+            Entries oldEntries = resolveCache.putIfAbsent(key, entries);
+            if (oldEntries != null) {
+                entries = oldEntries;
+            }
+        }
+
+        final long ttl;
+        Entry old = entries.add(e);
+        if (old != null) {
+            ScheduledFuture<?> future = old.expirationFuture;
+            if (future != null) {
+                ttl = Math.min(originalTtl, future.getDelay(TimeUnit.SECONDS));
```

Comment:
I think we should just port over the changes that we did in https://github.com/netty/netty/pull/8187 once it is merged. This will also "resolve" the comment.

Reply:
@Scottmitch done via baa3fd31638215828cb80af457eed1974a2fd209. If we adjust  #8187 I will bring over the changes.